Library/Homebrew/keg_relocate.rb

```diff
@@ -163,6 +167,45 @@ def replace_text_in_files(relocation, files: nil)
     changed_files
   end
 
+  def replace_build_prefix_in_file(old_cellar, new_prefix, new_cellar, file)
+    old_prefix = Pathname(old_cellar).parent.to_s
+
+    old_prefix_optdir = "#{old_prefix}/opt"
+    old_prefix_etcdir = "#{old_prefix}/etc"
+    old_prefix_vardir = "#{old_prefix}/var"
+    old_prefix_vim_sharedir = "#{old_prefix}/share/vim"
```

Comment:
This seems like a weird addition, why this/these specifically?

I think it'd be good to have these directories as a constant somewhere (perhaps reusing an existing constant) and loop over them?

Reply:
I took all of these from https://github.com/Homebrew/brew/blob/d92c544b5c441afe4607dac865ce8c2e17845bf0/Library/Homebrew/dev-cmd/bottle.rb#L479-L490. 
There doesn't seem to be a constant for this defined anywhere, so maybe we need one.  I debated whether we should just replace any occurrence of the old `HOMEBREW_PREFIX`, but given that we don't just scan for that when determining relocation, I'm assuming there was a reason for choosing those directories.  We might need to dig through commit histories to know when and why those specific directories were added.

Reply:
We should probably extract all of these into a separate method then if we want to relocate only those paths.

Reply:
> We should probably extract all of these into a separate method then if we want to relocate only those paths.

Agreed 👍🏻 

> We might need to dig through commit histories to know when and why those specific directories were added.

They are an incomplete subset covered by https://github.com/Homebrew/brew/blob/5fb34c8ef2ba88bfeb0bf19472ef83c90edd8464/Library/Homebrew/keg.rb#L413-L475

Relatedly, `etc` in a formula returns `HOMEBREW_PREFIX/"etc"` (same with `var`) but `bin` points returns `prefix/"bin"` (i.e. in the keg).

As a result, there's only a subset of things we need to relocate that reference `HOMEBREW_PREFIX` but not the keg or `opt`. It's probably worth trying to relocate them all at bottle time, though.

Reply:
That's very helpful information.  I was thinking about using the Relocation framework though I'll have to refactor the `gsub` part somewhat to go though each element of the Relocation object.  

Before going down that path, I think it's worth asking a more basic question: could we just replace all instances of `HOMEBREW_PREFIX`?  When going back over our codebase and Anaconda's I'm wondering if we were being conservative about which prefix subdirectories would make us consider a binary non-relocatable to avoid making binaries non-relocatable if they referenced include dirs etc. that aren't needed at runtime.

In Anaconda they just replace the build prefix regardless of whether the string it's in is needed at runtime. I can't really think of any reason we'd need to keep any references to the old build prefix in the binary.  Furthermore when the second part of this is fully implemented, all non-relocatable packages will be build with a longer placeholder prefix that nobody is actually using as real install prefix.  So maybe we could simplify this substantially by just replacing all references to the prefix?

Reply:
> Before going down that path, I think it's worth asking a more basic question: could we just replace all instances of HOMEBREW_PREFIX? 

No. If you're in `/usr/local` a lot of stuff refers to `/usr/local` in a way that doesn't relate to it being the Homebrew prefix e.g. CMake will do lookups for files in `/usr/local` regardless of where you've installed it with Homebrew. This doesn't apply to e.g. `/opt/homebrew`.

Reply:
This conversation is being continued below so I'm marking this part of it as resolved to consolidate things.